source/common/event/file_event_impl.cc

```diff
@@ -117,19 +128,60 @@ void FileEventImpl::setEnabled(uint32_t events) {
     injected_activation_events_ = 0;
     activation_cb_->cancel();
   }
+  updateEvents(events);
+}
 
-  auto* base = event_get_base(&raw_event_);
-  event_del(&raw_event_);
-  assignEvents(events, base);
-  event_add(&raw_event_, nullptr);
+void FileEventImpl::unregisterEventIfEmulatedEdge(uint32_t event) {
+  // This constexpr if allows the compiler to optimize away the function on POSIX
+  if constexpr (PlatformDefaultTriggerType == FileTriggerType::EmulatedEdge) {
+    ASSERT((event & (FileReadyType::Read | FileReadyType::Write)) == event);
+    if (trigger_ == FileTriggerType::EmulatedEdge) {
+      auto new_event_mask = enabled_events_ & ~event;
+      updateEvents(new_event_mask);
+    }
+  }
+}
+
+void FileEventImpl::registerEventIfEmulatedEdge(uint32_t event) {
+  // This constexpr if allows the compiler to optimize away the function on POSIX
+  if constexpr (PlatformDefaultTriggerType == FileTriggerType::EmulatedEdge) {
+    ASSERT((event & (FileReadyType::Read | FileReadyType::Write)) == event);
+    if (trigger_ == FileTriggerType::EmulatedEdge) {
+      auto new_event_mask = enabled_events_ | event;
+      if (event & FileReadyType::Read && (enabled_events_ & FileReadyType::Closed)) {
+        // We never ask for both early close and read at the same time.
+        new_event_mask = new_event_mask & ~FileReadyType::Read;
+      }
+      if (event == 0) {
+        return;
+      }
+      updateEvents(new_event_mask);
+    }
+  }
 }
 
 void FileEventImpl::mergeInjectedEventsAndRunCb(uint32_t events) {
   if (activate_fd_events_next_event_loop_ && injected_activation_events_ != 0) {
+    if (events & FileReadyType::Closed &&
+        activate_fd_events_next_event_loop_ & FileReadyType::Read) {
+      // We never ask for both early close and read at the same time. If close is requested
+      // keep that instead.
+      injected_activation_events_ = injected_activation_events_ & ~FileReadyType::Read;
+    }
+
     events |= injected_activation_events_;
     injected_activation_events_ = 0;
     activation_cb_->cancel();
   }
+
+  // TODO(davinci26): This can be optimized further in (w)epoll backends using the `EPOLLONESHOT`
```

Comment:
Please protect with:
if constexpr ( PlatformDefaultTriggerType == FileTriggerType::EmulatedEdge) {

Reply:
since `unregisterEventIfEmulatedEdge` is protected with `if constexpr ( PlatformDefaultTriggerType == FileTriggerType::EmulatedEdge)` this gets optimized away from the compiler.

Reply:
So it is implicitly protected

Reply:
Does the compiler remove the evaluation of the if predicate and the call to the now empty virtual function?  I can see compilers de-virtualizing the call to registerEventIfEmulatedEdge in this file.  I'm less sure they can prove that evaluation of the if predicate has no side effects and omit the virtual calls in source/common/network/io_socket_handle_impl.cc in methods like readv.

Reply:
verified it on godbolt that it cannot be optimized away if it is a virtual method. Added more aggressive guards.